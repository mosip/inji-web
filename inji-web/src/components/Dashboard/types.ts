import { Location } from "react-router-dom";

export type DropdownItem = {
    label: string;
    onClick: () => void;
    textColor: string;
    key: string;
};

export type SidebarItemType = {
    icon: React.ReactNode;
    text: string;
    path: string;
    key: string;
};

export type SidebarItemProps = {
    icon: React.ReactNode;
    text: string;
    path: string;
    isActive: boolean;
    isCollapsed: boolean;
};

export type SearchCredentialProps = {
    issuerContainerBorderRadius?: string;
}

export type FAQPageProps = {
    backUrl?: string;
    withHome?: boolean;
};

export type CredentialTypesPageProps = {
    backUrl?: string;
};

export type InfoFieldProps = {
    label?: string;
    value?: string;
};

export type CollapseButtonProps = {
    isCollapsed: boolean;
    onClick: () => void;
    className?: string;
};

export type FooterProps = {
    footerRef: React.RefObject<HTMLDivElement>;
};

export type HeaderProps = {
    headerRef: React.RefObject<HTMLDivElement>;
};

export type DashboardHeaderProps = {
    headerRef: React.RefObject<HTMLDivElement>;
    headerHeight: number;
};

export type SideBarSvgIconProps = {
    outline: string;
    navUrl: string;
    location: Location;
}

export type User = {
    displayName: string;
    profilePictureUrl: string;
    email:string;
};

export type ErrorType = {
    errorCode: string;
    errorMessage: string;
};

export type UserContextType = {
    user: User | null;
    walletId: string | null;
    error: ErrorType | null;
    fetchUserProfile: () => Promise<{user: User; walletId: string}>;
    saveUser: (user: User) => void;
    removeUser: () => void;
    isLoading: boolean;
};
