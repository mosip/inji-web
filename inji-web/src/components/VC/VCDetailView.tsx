import {WalletCredential} from "../../types/data";
import {Modal} from "../../modals/Modal";
import {PDFViewer} from "../Preview/PDFViewer";
import React from "react";
import {ResponsiveIconButtonWithText} from "../Common/Buttons/ResponsiveIconButtonWithText";
import {DownloadIcon} from "../Common/Icons/DownloadIcon";
import {useTranslation} from "react-i18next";


export function VCDetailView(props: Readonly<{
    previewContent: string,
    onClose: () => void,
    onDownload: () => Promise<void>,
    credential: WalletCredential
}>) {
    const {t} = useTranslation('StoredCards',{
        keyPrefix: "cardView"
    })

    return <Modal isOpen={!!props.previewContent}
                  onClose={props.onClose}
                  action={<ResponsiveIconButtonWithText text={t('download')}
                                                        icon={<DownloadIcon
                                                            testId={"icon-download"}
                                                        />}
                                                        testId={"download"}
                                                        onClick={props.onDownload}
                  />}
                  title={props.credential.credentialTypeDisplayName}
    >
        <PDFViewer
            previewContent={props.previewContent}
        />
    </Modal>;
}