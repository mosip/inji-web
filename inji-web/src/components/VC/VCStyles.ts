export const VCStyles = {
    cardView : {
        container: "bg-iw-tileBackground grid grid-cols-[2.5fr_0.5fr] items-center shadow hover:shadow-lg hover:scale-105 hover:shadow-iw-selectedShadow p-4 m-2 rounded-md cursor-pointer",
        contentsContainer: "flex items-center overflow-scroll gap-6 sm:gap-4",
        logo: "w-16 h-16 sm:w-20 sm:h-20",
        credentialName: "text-sm font-semibold text-iw-title",
        actionsContainer: "flex flex-row items-center flex-shrink-0",
        downloadIcon: "h-25 w-25",
        menuIcon: "m-1"
    }
}