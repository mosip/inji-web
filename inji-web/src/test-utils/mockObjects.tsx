import { IssuerConfigurationObject } from '../types/data';
import { LandingPageWrapperProps } from '../components/Common/LandingPageWrapper';
import { NavBarProps } from '../types/components';
export const mockIssuerDisplayArrayObject = {
    name: "Name",
    language: "en",
    locale: "en",
    logo: {
        url: "https://url.com",
        alt_text: "alt text of the url"
    },
    title: "Title",
    description: "Description",
};

export const mockCredentialTypeDisplayArrayObject = {
    name: "Name",
    locale: "en",
    logo: "https://url.com",
};

export const mockCredentials: IssuerConfigurationObject = {
    credentials_supported: [
        {
            name: "InsuranceCredential",
            scope: "mosip_vc_ldp",
            display: [
                {
                    name: "Health Insurance",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        },
        {
            name: "AnotherCredential",
            scope: "mosip_ldp_vc",
            display: [
                {
                    name: "Another Credential",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        }
    ],
    "authorization_endpoint": "https://env.net/authorize",
    "grant_types_supported": ["authorization_code"]
};


export const mockIssuerObject = {
        name: 'Test Issuer',
        desc: 'Test Description',
        protocol: 'OTP' as 'OTP', // Explicitly set the type to 'OTP' or 'OpenId4VCI'
        issuer_id: 'test-issuer',
        authorization_endpoint: 'https://auth.test.com',
        credentials_endpoint: 'https://credentials.test.com',
        display: [{
            name: "Name",
            language: "en",
            locale: "en",
            logo: {
                url: "https://url.com",
                alt_text: "alt text of the url"
            },
            title: "Title",
            description: "Description",
        }],
        client_id: 'test-client-id',
        redirect_uri:'test-redirect-uri',
        token_endpoint:'test-token_endpoint',
        proxy_token_endpoint:'test-proxy_token_endpoint',
        client_alias:'',
        ovp_qr_enabled: true,
        scopes_supported: ['scope1', 'scope2']
}

export const mockIssuerObjectList =  [
    {
        name: 'Issuer 1',
        desc: 'Description 1',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer1',
        authorization_endpoint: 'https://issuer1.com/auth',
        credentials_endpoint: 'https://issuer1.com/credentials',
        display: [
            {
                name: 'Issuer 1',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer1.com/logo.png', alt_text: 'Issuer 1 Logo' },
                title: 'Issuer 1 Title',
                description: 'Issuer 1 Description',
            },
        ],
        client_id: 'client1',
        redirect_uri: 'https://issuer1.com/redirect',
        token_endpoint: 'https://issuer1.com/token',
        proxy_token_endpoint: 'https://issuer1.com/proxy-token',
        client_alias: 'client1-alias',
        ovp_qr_enabled: true,
        scopes_supported: ['openid', 'profile'],
    },
    {
        name: 'Issuer 2',
        desc: 'Description 2',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer2',
        authorization_endpoint: 'https://issuer2.com/auth',
        credentials_endpoint: 'https://issuer2.com/credentials',
        display: [
            {
                name: 'Issuer 2',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer2.com/logo.png', alt_text: 'Issuer 2 Logo' },
                title: 'Issuer 2 Title',
                description: 'Issuer 2 Description',
            },
        ],
        client_id: 'client2',
        redirect_uri: 'https://issuer2.com/redirect',
        token_endpoint: 'https://issuer2.com/token',
        proxy_token_endpoint: 'https://issuer2.com/proxy-token',
        client_alias: 'client2-alias',
        ovp_qr_enabled: false,
        scopes_supported: ['openid', 'profile'],
    },
]

export const mockLandingPageWrapperProps: LandingPageWrapperProps = {
    icon: <div data-testid="Test-Icon">Icon</div>,
    title: "Test Title",
    subTitle: "Test SubTitle",
    gotoHome: true,
};

export const mockNavBarProps: NavBarProps = {
    title: 'Test Title',
    link: '/test-link',
    search: true,
};