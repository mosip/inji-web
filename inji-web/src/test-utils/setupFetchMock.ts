export const fetchMock = jest.fn();

beforeAll(() => {
    global.fetch = fetchMock;
});

beforeEach(() => {
    fetchMock.mockReset();
});
