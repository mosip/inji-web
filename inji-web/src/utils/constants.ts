export const KEYS = {
    USER: 'user',
    WALLET_ID: 'walletId',
};

// This constant exposes the pages used in the application. This will be used only by the Router component.
export const Pages = {
    ROOT: "/",
    USER: "user",
    HOME: "home",
    ISSUERS: "issuers",
    ISSUER_TEMPLATE: "issuers/:issuerId",
    CREDENTIALS: "credentials",
    PROFILE: "profile",
    FAQ: "faq",
    REDIRECT: "redirect",
    RESET_PASSCODE: "reset-passcode",
    PASSCODE: "passcode",
    AUTHORIZE: "authorize"
}

// This constant exposes the routes of the application which can be used across the component to avoid hardcoded navigation urls
export const ROUTES = {
    ROOT: Pages.ROOT,
    USER: `/${Pages.USER}`,
    PASSCODE: `/${Pages.USER}/${Pages.PASSCODE}`,
    USER_HOME: `/${Pages.USER}/${Pages.HOME}`,
    USER_ISSUERS: `/${Pages.USER}/${Pages.ISSUERS}`,
    ISSUERS: `/${Pages.ISSUERS}`,
    USER_ISSUER: (id: string) => `/${Pages.USER}/${Pages.ISSUERS}/${id}`,
    ISSUER: (id: string) => `/${Pages.ISSUERS}/${id}`,
    CREDENTIALS: `/${Pages.USER}/${Pages.CREDENTIALS}`,
    PROFILE: `/${Pages.USER}/${Pages.PROFILE}`,
    FAQ: `/${Pages.FAQ}`,
    USER_FAQ: `/${Pages.USER}/${Pages.FAQ}`,
    USER_RESET_PASSCODE: `/${Pages.USER}/${Pages.RESET_PASSCODE}`,
} as const;

export const HTTP_STATUS_CODES = {
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    INTERNAL_SERVER_ERROR: 500,
    SERVICE_UNAVAILABLE: 503,
}

export const PDF_WORKER_URL = "https://unpkg.com/pdfjs-dist@3.11.174/build/pdf.worker.min.js";
