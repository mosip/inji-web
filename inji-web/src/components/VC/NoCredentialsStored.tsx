import {useTranslation} from "react-i18next";
import {navigateToDashboardHome} from "../../pages/Dashboard/utils";
import React from "react";

export function NoCredentialsStored() {
    const {t} = useTranslation('Dashboard');

    return <div
        data-testid={"Blank-Document"}
        className="bg-white rounded-lg shadow-iw-emptyDocuments p-6 sm:p-8 md:p-10 flex flex-col items-center justify-center min-h-[500px] my-4 sm:my-8 md:my-16"
    >
        <svg
            data-testid={"Document-Icon"}
            width="88"
            height="88"
            viewBox="0 0 68 88"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            className="mb-8 sm:mb-12 md:mb-20"
        >
            <path
                d="M17.847 88C15.611 88 13.744 87.246 12.246 85.739C10.749 84.231 10 82.353 10 80.102V7.898C10 5.647 10.749 3.769 12.246 2.261C13.744 0.754 15.611 0 17.847 0H56.143L78 22V80.102C78 82.353 77.251 84.231 75.754 85.739C74.256 87.246 72.389 88 70.153 88H17.847ZM53.714 24.444H73.143L53.714 4.889V24.444ZM27 70.889H61V66H27V70.889ZM27 51.333H61V46.444H27V51.333Z"
                fill="#DEDEDE"
            />
        </svg>
        <h2
            data-testid={"No-Credentials-Title"}
            className="text-xl text-center sm:text-2xl font-medium text-gray-800 mb-2"
        >
            {t('StoredCredentials.emptyScreen.title')}
        </h2>
        <div className="block sm:hidden mt-6">
            <button
                data-testid={"Blank-Document-Add-Credential"}
                onClick={navigateToDashboardHome}
                className="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-6 py-2 rounded-lg text-sm font-semibold shadow-sm transition-shadow"
            >
                {t('StoredCredentials.header.addCredential')}
            </button>
        </div>
    </div>;
}