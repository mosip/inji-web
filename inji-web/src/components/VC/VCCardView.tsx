import {WalletCredential} from "../../types/data";
import {RxDotsHorizontal} from "react-icons/rx";
import {VCStyles} from "./VCStyles";
import React, {useState} from "react";
import DownloadIcon from "../../assets/Download.svg"
import {Clickable} from "../Common/Clickable";
import {api} from "../../utils/api";
import {downloadCredentialPDF} from "../../utils/misc";
import {useSelector} from "react-redux";
import {RootState} from "../../types/redux";
import {toast} from "react-toastify";
import {Viewer, Worker} from "@react-pdf-viewer/core";
import {Modal} from "../../modals/Modal";
import {SolidButton} from "../Common/Buttons/SolidButton";

function VCDetailView(props: { previewContent: string, onClick: () => void }) {
    return (
        <Worker
            workerUrl={`https://unpkg.com/pdfjs-dist@3.11.174/build/pdf.worker.min.js`}
        >
            <Viewer fileUrl={props.previewContent}/>
        </Worker>
    )
}


export function VCCardView(props: Readonly<{
    credential: WalletCredential
}>) {
    const language = useSelector((state: RootState) => state.common.language);
    const [error, setError] = useState<string>()
    const [previewContent, setPreviewContent] = useState<string>("");


    if (error) {
        toast.error("Download failed. Please try again later.");
    }

    const preview = async () => {
        console.log("Fetching credential preview for:", props.credential.credentialId);
        try {
            const response = await fetch(
                api.fetchWalletCredentialPreview.url(props.credential.credentialId),
                {
                    method:
                        api.fetchWalletCredentialPreview.methodType === 0
                            ? "GET"
                            : "POST",
                    headers: api.fetchWalletCredentialPreview.headers(language),
                    credentials: api.fetchWalletCredentialPreview.credentials
                }
            );
            //
            const pdfContent = await response.blob();

            const pdfUrl = URL.createObjectURL(pdfContent);
            setPreviewContent(pdfUrl)
        } catch (error) {
            console.error("Failed to download credential PDF:", error);
            setError("downloadError");
        }
    }

    const download = async (mouseEvent : React.MouseEvent) => {
        mouseEvent.stopPropagation()
        console.log("Downloading credential PDF for:", props.credential.credentialId);
        try {
            const response = await fetch(
                api.fetchWalletCredentialPreview.url(props.credential.credentialId),
                {
                    method:
                        api.fetchWalletCredentialPreview.methodType === 0
                            ? "GET"
                            : "POST",
                    headers: api.fetchWalletCredentialPreview.headers(language),
                    credentials: api.fetchWalletCredentialPreview.credentials
                }
            );
            //
            const pdfContent = await response.blob();

            const disposition = response.headers.get("Content-Disposition");
            const fileNameMatch = /filename="(.+)"/.exec(disposition ?? "");
            const fileName = fileNameMatch?.[1] || "download.pdf";

            await downloadCredentialPDF(pdfContent, fileName);
        } catch (error) {
            console.error("Failed to download credential PDF:", error);
            setError("downloadError");
        }
    }

    const clearPreview = () => {
        setPreviewContent("");
    }

    return (
        <Clickable onClick={preview} testId={"vc-card-view"} className={VCStyles.cardView.container}>
            <Modal isOpen={!!previewContent}
                   onClose={clearPreview}
                   action={<SolidButton testId={"btn-download"} title={"download"} onClick={download}/>}
                   title={props.credential.credentialTypeDisplayName}
            >
                <VCDetailView
                    previewContent={previewContent}
                    onClick={clearPreview}
                />
            </Modal>

            <div className={"flex items-center gap-6"}>
                <img
                    data-testid="issuer-logo"
                    src={props.credential.credentialTypeLogo}
                    alt="Credential Type Logo"
                    className={VCStyles.cardView.logo}
                />
                <span
                    className={VCStyles.cardView.credentialName}
                    data-testid="credential-type-display-name"
                >
                {props.credential.credentialTypeDisplayName}
            </span>
            </div>
            <div className={"flex flex-row gap-2"}>
                <img
                    onClick={download}
                    data-testid={"icon-download"}
                    src={DownloadIcon}
                    alt={"icon-download"}
                    className={"h-25 w-25"}
                />
                <RxDotsHorizontal size={20} color={"#707070"}/>
            </div>
        </Clickable>
    );
}