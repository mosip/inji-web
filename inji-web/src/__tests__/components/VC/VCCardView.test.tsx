import React from 'react';
import {fireEvent, screen} from '@testing-library/react';
import {WalletCredential} from '../../../types/data';
import {VCCardView} from '../../../components/VC/VCCardView';
import {
    mockLocalStorage,
    mockUseSelector,
    renderWithProvider,
    setMockUseSelectorState
} from "../../../test-utils/mockUtils";
import {fetchMock} from "../../../test-utils/setupFetchMock";
import {KEYS} from "../../../utils/constants.ts";
import {mockVerifiableCredentials} from "../../../test-utils/mockObjects.tsx";

describe('VCCardView Component', () => {
    const mockCredential: WalletCredential = mockVerifiableCredentials[0]
    let localStorageMock


    beforeEach(() => {
        jest.clearAllMocks();

        mockUseSelector();
        setMockUseSelectorState({
            common: {
                language: 'en',
            },
        });

        localStorageMock = mockLocalStorage();

        localStorageMock.setItem(KEYS.WALLET_ID, "faa0e18f-0935-4fab-8ab3-0c546c0ca714")
    });

    it('should match snapshot', () => {
        const {container} = renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        expect(container).toMatchSnapshot();
    });

    it('should render with correct credential information', () => {
        renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        expect(screen.getByTestId("vc-card-view")).toBeInTheDocument();
        const logo = screen.getByTestId('issuer-logo');
        const name = screen.getByTestId('credential-type-display-name');
        expect(logo).toBeInTheDocument();
        expect(logo).toHaveAttribute('src', 'logo1.png');
        expect(logo).toHaveAttribute('alt', 'Credential Type Logo');

        expect(name).toBeInTheDocument();
        expect(name).toHaveTextContent('Drivers License');
    });

    it('should call download api when clicked', () => {
        fetchMock.mockResolvedValueOnce({
            ok: true,
            blob: async () => new Blob(),
            headers: {
                get: () => 'attachment; filename="credential.pdf"',
            },
        });
        renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        const card = screen.getByRole('menuitem');
        fireEvent.click(card);

        expect(fetchMock).toHaveBeenCalledTimes(1);
    });

    it('should call download api when pressing enter key', () => {
        renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        const card = screen.getByRole('menuitem');
        fireEvent.keyDown(card, {key: 'Enter'});


    });

    it('should not call download api when pressing key other than enter key', () => {
        renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        const card = screen.getByRole('menuitem');
        fireEvent.keyDown(card, {key: '1'});

        expect(fetchMock).not.toBeCalled()
    });

    it('should call download api when download icon is clicked', () => {
        fetchMock.mockResolvedValueOnce({
            ok: true,
            blob: async () => new Blob(),
            headers: {
                get: () => 'attachment; filename="credential.pdf"',
            },
        });
        renderWithProvider(
            <VCCardView
                credential={mockCredential}
            />
        );

        const downloadIcon = screen.getByTestId('icon-download');
        fireEvent.click(downloadIcon);

        expect(fetchMock).toHaveBeenCalledTimes(1);
        expect(fetchMock).toHaveBeenCalledWith(
            expect.stringContaining("wallets/faa0e18f-0935-4fab-8ab3-0c546c0ca714/credentials/cred-1?action=download"),
            expect.objectContaining({
                "credentials": "include",
                "headers": {"Accept": "application/pdf", "Accept-Language": "en", "Content-Type": "application/json"},
                "method": "GET"
            })
        );
    })
});