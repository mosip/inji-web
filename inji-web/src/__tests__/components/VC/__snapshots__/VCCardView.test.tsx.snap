// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCCardView Component should match snapshot 1`] = `
<div>
  <div
    class="bg-iw-tileBackground grid grid-cols-[auto_auto] items-center shadow hover:shadow-lg hover:scale-105 hover:shadow-iw-selectedShadow p-4 m-2 rounded-md cursor-pointer"
    data-testid="vc-card-view"
    role="menuitem"
    tabindex="0"
  >
    <div
      class="flex items-center gap-6"
    >
      <img
        alt="Credential Type Logo"
        class="w-20 h-20"
        data-testid="issuer-logo"
        src="logo1.png"
      />
      <span
        class="text-sm font-semibold text-iw-title"
        data-testid="credential-type-display-name"
      >
        Drivers License
      </span>
    </div>
    <div
      class="flex flex-row gap-2"
    >
      <img
        alt="icon-download"
        class="h-25 w-25"
        data-testid="icon-download"
        src="Download.svg"
      />
      <svg
        color="#707070"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(112, 112, 112);"
        viewBox="0 0 15 15"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3.625 7.5C3.625 8.12132 3.12132 8.625 2.5 8.625C1.87868 8.625 1.375 8.12132 1.375 7.5C1.375 6.87868 1.87868 6.375 2.5 6.375C3.12132 6.375 3.625 6.87868 3.625 7.5ZM8.625 7.5C8.625 8.12132 8.12132 8.625 7.5 8.625C6.87868 8.625 6.375 8.12132 6.375 7.5C6.375 6.87868 6.87868 6.375 7.5 6.375C8.12132 6.375 8.625 6.87868 8.625 7.5ZM12.5 8.625C13.1213 8.625 13.625 8.12132 13.625 7.5C13.625 6.87868 13.1213 6.375 12.5 6.375C11.8787 6.375 11.375 6.87868 11.375 7.5C11.375 8.12132 11.8787 8.625 12.5 8.625Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
